File: datasets/Python/Chatgpt/DataStructures/medium4.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.count = 0  # Number of words passing through this node

class Contacts:
    def __init__(self):
        self.root = TrieNode()

    def add(self, name):
        node = self.root
        for ch in name:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
            node.count += 1

    def find(self, partial):
        node = self.root
        for ch in partial:
            if ch not in node.children:
                return 0
            node = node.children[ch]
        return node.count

# MAIN FUNCTION
def main():
    n = int(input("Enter number of operations: "))
    queries = [input() for _ in range(n)]

    contacts = Contacts()
    result = []

    for query in queries:
        op, *args = query.strip().split()
        if op == "add":
            contacts.add(args[0])
        elif op == "find":
            result.append(contacts.find(args[0]))

    print("Output:")
    for val in result:
        print(val)

main()

File: datasets/C++/Chatgpt/DataStructures/medium4.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>
using namespace std;

struct TrieNode {
    unordered_map<char, TrieNode*> children;
    int count = 0;
};

class Contacts {
    TrieNode* root;

public:
    Contacts() {
        root = new TrieNode();
    }

    void add(const string& name) {
        TrieNode* node = root;
        for (char c : name) {
            if (node->children.count(c) == 0)
                node->children[c] = new TrieNode();
            node = node->children[c];
            node->count++;
        }
    }

    int find(const string& partial) {
        TrieNode* node = root;
        for (char c : partial) {
            if (node->children.count(c) == 0)
                return 0;
            node = node->children[c];
        }
        return node->count;
    }
};

int main() {
    int n;
    cin >> n;
    cin.ignore();
    Contacts contacts;
    vector<int> result;

    for (int i = 0; i < n; ++i) {
        string op, arg;
        cin >> op >> arg;
        if (op == "add") {
            contacts.add(arg);
        } else if (op == "find") {
            result.push_back(contacts.find(arg));
        }
    }

    for (int val : result) {
        cout << val << endl;
    }

    return 0;
}
